### PR TITLE
*: adjust delta schema count and add metrics

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
The original size of `deltaSchemaInfos` is 100. 
Executing some SQLs may take a long time. At the same time, the TiDB cluster may do a lot of DDLs, which means we change a lot of schema versions. And the schema version changed in it may exceed 100. Then the SQLs will return the error of `Information schema is changed.`.

If the DDL owner TiDB and the receiving request TiDB are on the same TiDB, it takes 50 + ms to update each mode version. 
If the DDL owner TiDB and the receiving request TiDB aren't on the same TiDB, it takes 500 + ms to update each mode version. 

The default of `MaxTxnTimeUse` is 590s. So I set the size of `deltaSchemaInfos` to 1024.

### What is changed and how it works?

-  Add a variable of `TiDBMaxDeltaSchemaCount` to update the max size of `deltaSchemaInfos `  . Besides, add a function of `merge`

- Add some metrics.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by winkyao** [comment]
@zimulala Please fix ci

**Comment by zimulala** [comment]
PTAL @winkyao @jackysp

File: session/session.go [hunk 1]
```diff
@@ -1770,6 +1770,7 @@ var builtinGlobalVariable = []string{
 	variable.TiDBExpensiveQueryTimeThreshold,
 	variable.TiDBEnableNoopFuncs,
 	variable.TiDBEnableIndexMerge,
+	variable.TiDBMaxDeltaSchemaCount,
```
Comment (jackysp) [inline session/session.go]: 屏幕快照 2019-08-22 下午1 58 48
Permission changed?

File: domain/schema_validator.go [hunk 2]
```diff
@@ -65,14 +67,15 @@ type schemaValidator struct {
 	latestSchemaExpire time.Time
 	// deltaSchemaInfos is a queue that maintain the history of changes.
 	deltaSchemaInfos []deltaSchemaInfo
+	notMergeCnt      int
```
Comment (jackysp) [inline domain/schema_validator.go]: Can not quite understand what is it used for. Plz add some comments.

**Comment by zimulala** [comment]
PTAL @jackysp @winkyao

**Comment by zimulala** [comment]
PTAL @tiancaiamao  @jackysp @winkyao

**Comment by zimulala** [comment]
PTAL @jackysp done.

File: domain/schema_validator.go [hunk 3]
```diff
@@ -209,7 +220,35 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
 
 func (s *schemaValidator) enqueue(schemaVersion int64, relatedTableIDs []int64) {
 	s.deltaSchemaInfos = append(s.deltaSchemaInfos, deltaSchemaInfo{schemaVersion, relatedTableIDs})
-	if len(s.deltaSchemaInfos) > maxNumberOfDiffsToLoad {
+	s.notMergeCnt++
+
+	maxCnt := int(variable.GetMaxDetalSchemaCount())
+	if len(s.deltaSchemaInfos) > maxCnt && s.notMergeCnt > maxCnt/2 {
+		s.merge()
+		s.notMergeCnt = 1
```
Comment (jackysp) [inline domain/schema_validator.go]: Why set to 1 here? I think we don't know how many `deltaSchemaInfos` merged here.

Comment (zimulala) [reply domain/schema_validator.go]: We call `merge` before setting the value. So I think that the previous `deltaSchemaInfos` have been merged.

**Comment by zimulala** [comment]
PTAL @jackysp @tiancaiamao @winkyao

**Comment by zimulala** [comment]
PTAL @tiancaiamao @winkyao

File: domain/schema_validator.go [hunk 4]
```diff
@@ -209,7 +220,35 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
...
+	}
+	for i := 0; i < len(a); i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *schemaValidator) merge() {
```
Comment (bb7133) [inline domain/schema_validator.go]: Is it possible to try to merge delta info just at the first time of `enque`?
```
s.deltaSchemaInfos = append(s.deltaSchemaInfos, deltaSchemaInfo{schemaVersion, relatedTableIDs})
```

File: domain/schema_validator.go [hunk 5]
```diff
@@ -209,7 +220,35 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
...
+	if len(s.deltaSchemaInfos) > maxCnt && s.notMergeCnt > maxCnt/2 {
+		s.merge()
+		s.notMergeCnt = 1
+	}
+	if len(s.deltaSchemaInfos) > maxCnt {
 		s.deltaSchemaInfos = s.deltaSchemaInfos[1:]
 	}
 }
+
+func equal(a, b []int64) bool {
```
Comment (tiancaiamao) [inline domain/schema_validator.go]: https://godoc.org/bytes#Equal

**Comment by zz-jason** [comment]
@zimulala any update?

File: domain/schema_validator.go (hunk 5 above)
Comment (zimulala) [reply domain/schema_validator.go]: @tiancaiamao There are different types. The argument types of `equal` are `[]int64`.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
@jackysp

File: domain/schema_validator.go (hunk 4 above)
Comment (zimulala) [reply domain/schema_validator.go]: Good idea. Done

**Comment by zimulala** [comment]
PTAL @bb7133 
@jackysp @tiancaiamao  Do you need to review again?

File: domain/schema_validator.go [hunk 6]
```diff
@@ -208,8 +215,51 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
...
+	}
+
+	if len(s.deltaSchemaInfos) > maxCnt {
+		logutil.BgLogger().Info("the schema validator enqueue, queue is too long",
+			zap.Int("delta max count", maxCnt), zap.Int64("remove schema version", s.deltaSchemaInfos[0].schemaVersion))
 		s.deltaSchemaInfos = s.deltaSchemaInfos[1:]
 	}
 }
+
+func unorderedEqual(a, b []int64) bool {
```
Comment (AilinKid) [inline domain/schema_validator.go]: When `b` is a superset of `a` is ok, so the name of the function should be changed.

Comment (tiancaiamao) [reply domain/schema_validator.go]: This function is not correct. @AilinKid @zimulala

Comment (zimulala) [reply domain/schema_validator.go]: Done.

**Comment by zimulala** [comment]
PTAL @AilinKid @tiancaiamao

File: domain/schema_validator.go [hunk 7]
```diff
@@ -208,8 +215,54 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
...
+	}
+
+	if len(s.deltaSchemaInfos) > maxCnt {
+		logutil.BgLogger().Info("the schema validator enqueue, queue is too long",
+			zap.Int("delta max count", maxCnt), zap.Int64("remove schema version", s.deltaSchemaInfos[0].schemaVersion))
 		s.deltaSchemaInfos = s.deltaSchemaInfos[1:]
 	}
 }
+
+type ids []int64
```
Comment (tiancaiamao) [inline domain/schema_validator.go]: Defined but not used?

Comment (tiancaiamao) [reply domain/schema_validator.go]: Oh, used in line 248

**Comment by tiancaiamao** [comment]
LGTM

**Comment by AilinKid** [comment]
LGTM

**Pull request merged**